enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,355 @@
+	// +kubebuilder:validation:Maximum:=43
+	// +kubebuilder:validation:Minimum:=0
+	IcmpType int8 `json:"icmpType" protobuf:"varint,2,opt,name=icmpType"`
+
+	// IcmpCode define ICMP Code ID (RFC 792).
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Maximum:=16
+	// +kubebuilder:validation:Minimum:=0
+	IcmpCode int8 `json:"icmpCode" protobuf:"varint,2,opt,name=icmpCode"`
+
+	// Protocol can be ICMP, ICMPv6, TCP, SCTP or UDP.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="ICMP";"ICMPV6";"TCP";"UDP";"SCTP"
+	Protocol IngressNodeFirewallRuleProtocolType `json:"protocol" protobuf:"bytes,2,opt,name=protocol,casttype=IngressNodeFirewallRuleProtocolType"`
+
+	// Action can be Allow or Deny.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="Allow";"Deny"
+	Action IngressNodeFirewallActionType `json:"action" protobuf:"bytes,3,opt,name=action,casttype=IngressNodeFirewallActionType"`
+}
+
+// ProtocolType defines the protocol types that are supported
+type IngressNodeFirewallRuleProtocolType string
+
+const (
+	// ProtocolTypeICMP refers to the ICMP protocol
+	ProtocolTypeICMP ProtocolType = "ICMP"
+
+	// ProtocolTypeICMPV6 refers to the ICMP protocol
+	ProtocolTypeICMPV6 ProtocolType = "ICMPV6"
+
+	// ProtocolTypeTCP refers to the TCP protocol
+	ProtocolTypeTCP ProtocolType = "TCP"
+
+	// ProtocolTypeUDP refers to the UDP protocol
+	ProtocolTypeUDP ProtocolType = "UDP"
+
+	// ProtocolTypeSCTP refers to the SCTP protocol
+	ProtocolTypeSCTP ProtocolType = "SCTP"
+)
+
+// IngressNodeFirewallActionType indicates whether an IngressNodeFirewallRule allows or denies traffic
+// +kubebuilder:validation:Pattern=`^Allow|Deny$`
+type IngressNodeFirewallActionType string
+
+const (
+	IngressNodeFirewallAllow IngressNodeFirewallActionType = "Allow"
+	DIngressNodeFirewallDeny IngressNodeFirewallActionType = "Deny"
+)
+
+// IngressNodeFirewallRules define ingress node firewall rule
+type IngressNodeFirewallRules struct {
+// FromCIDRS is A list of CIDR from which we apply node firewall rule
+FromCIDRs []IngressNodeFirewallFromCIDR `json:"fromCIDRs" protobuf:"bytes,1,opt,name=fromCIDRs"`
+// FirewallProtocolRules is A list of per protocol ingress node firewall rules
+FirewallProtocolRules []IngressNodeFirewallProtocolRule `json:"firewallProtocolRules" protobuf:"bytes,2,opt,names=firewallProtocolRules"`
+}
+
+// NodeIngressFirewallConfigConfigSpec consists of a node name which the private IP should be assigned to.
+// +k8s:openapi-gen=true
+type NodeIngressFirewallConfigSpec struct {
+	// A list of interfaces where the ingress firewall policy will be applied on.
+	// empty list indicates the firewall policy applied on all interfaces
+	// +kubebuilder:validation:Optional
+	// +optional
+	Interface []string `json:"interface" protobuf:"bytes,1,opt,name=interface"`
+
+	// A list of ingress firewall policy rules.
+	// empty list indicates no ingress firewall i.e allow all incoming traffic.
```

Comment:
(This one is fine, because that's actually consistent; if there are 0 rules in the list, then the firewall matches 0 kinds of traffic.)